### PR TITLE
Enable NonNullByDefault for lspe4.plugin

### PR DESCRIPTION
**Pull request by sebthom**
I'd propose the introduction of NonNullByDefault Eclipse annotation on package level for improved null safety.
The current code base is very inconsistent regarding null value handling and contains hidden null pointer issues that e.g. result in issues like this https://github.com/eclipse/lsp4e/issues/913

My goal is to progressively enable the NonNullByDefault on package level(s).

I have good experience in TM4E moving to NonNullByDefault. I found hidden bugs in the control flow and it now gives me increased confidence in working with the API and during refactorings.

@mickaelistria @rubenporras do you support this change? 

Reference: https://help.eclipse.org/latest/topic/org.eclipse.jdt.doc.user/tasks/task-using_null_annotations.htm

### CODE REVIEW

File: org.eclipse.lsp4e/build.properties [hunk 1]
```diff
@@ -10,3 +10,9 @@ bin.includes = META-INF/,\
                schema/,\
                OSGI-INF/
 src.includes = schema/
+
+# https://codeiseasy.wordpress.com/2013/03/08/tycho-and-jdt-null-analysis/
+# JDT Null Analysis for Eclipse
+additional.bundles = org.eclipse.jdt.annotation
```
Comment (rubenporras) [inline org.eclipse.lsp4e/build.properties:16]: the blogpost does not say why this preferable to having it in the MANIFEST with optional resolution. Do you know the reason?

**Comment by rubenporras** [comment]
> @mickaelistria @rubenporras do you support this change?

From my side, yes, I like the idea. I still need to go through the PR to have a better idea.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyView.java [hunk 2]
```diff
@@ -30,12 +33,12 @@
...
 	public static final String ID = "org.eclipse.lsp4e.callHierarchy.callHierarchyView"; //$NON-NLS-1$
 
-	protected TreeViewer treeViewer;
+	protected TreeViewer treeViewer = lazyNonNull();
 
 	private final CallHierarchyContentProvider contentProvider = new CallHierarchyContentProvider();
 
 	@Override
-	public void createPartControl(final Composite parent) {
+	public void createPartControl(final @NonNullByDefault({}) Composite parent) {
```
[the remark this answers is not in the document]
Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyView.java:41]: This disables the automatically applied NonNull annotation and makes the parameter unannotated. 

This is required if the base class itself has no annotations applied itself but the parameter effectively is NonNull. Alternatively one can apply the Nullable annotation instead but then a programmatic null check needs to be performed to satisfy the compiler.

Another alternative is to provide EEA files which can apply the correct annotations to the base class. I am currently working on this, as I would prefer to have proper annotations.

The PR is work in progress. My main interest is if you are supportive of such an undertaking.

**Comment by mickaelistria** [comment]
I'd be fine with @NonNullByDefault as much as possible. `null` is the worst thing in the Java language.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyView.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rubenporras) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyView.java:41]: I agree that providing EEA files which apply the correct annotations to the external class is the way to go. 

Yes, I support your effort to have a codebase that can be checked for NPEs.

File: org.eclipse.lsp4e/build.properties (hunk 1 above)
Comment (sebthom) [reply org.eclipse.lsp4e/build.properties:15]: My understanding is that MANIFEST entries with optional resolution are still evaluated during runtime but we really don't need that bundle during runtime. Dependencies in build.properties are strictly used during compilation/build.

**Comment by sebthom** [comment]
@rubenporras I am aware that this is a huge PR. Enabling annotation based null analysis affects everything, so I had to adjust all classes of the plugin. I hope no other PRs are merged before this one, as that would lead to substantial effort in resolving merge conflicts.

**Comment by rubenporras** [comment]
@sebthom , you just asked for the review at the beginning of my summer holidays ;) I will check it today.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServers.java [hunk 3]
```diff
@@ -172,7 +171,8 @@ public <T> CompletableFuture<Optional<T>> computeFirst(BiFunction<? super Langua
 				executeOnServers(queryLS)
 				.map(lsRequest -> {
 					CompletableFuture<T> populateFuture = lsRequest.thenApply(t -> {
-						if (!isEmpty(t)) { // some LS methods return null objects when they have nothing to report, and some return an empty List
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServers.java:175]: could we undo this part of the PR and keep the comment and the `isEmpty` method?

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyContentProvider.java [hunk 4]
```diff
@@ -33,29 +35,31 @@
...
+	public Object[] getElements(final @Nullable Object inputElement) {
 		if (rootItems != null) {
 			return rootItems.toArray();
 		}
 		return new Object[] { rootMessage };
 	}
 
 	@Override
-	public Object[] getChildren(final Object parentElement) {
+	public Object [] getChildren(final Object parentElement) {
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyContentProvider.java:62]:
```suggestion
	public Object[] getChildren(final Object parentElement) {
```
to make it consistent with the rest of the codebase

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyContentProvider.java [hunk 5]
```diff
@@ -86,42 +90,38 @@ public void inputChanged(final Viewer viewer, final Object oldInput, final Objec
...
 
-	private void initialise(final @NonNull IDocument document, final int offset) throws BadLocationException {
+	private void initialise(final IDocument document, final int offset) throws BadLocationException {
 		LanguageServerDocumentExecutor executor = LanguageServers.forDocument(document)
 				.withCapability(ServerCapabilities::getCallHierarchyProvider);
 		if (!executor.anyMatching()) {
 			handleRootError();
 			return;
 		}
-		CallHierarchyPrepareParams prepareParams = LSPEclipseUtils.toCallHierarchyPrepareParams(offset, document);
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyContentProvider.java:111]: To keep it consistent with the rest of the elements, I would leave it in `LSPEclipseUtils`

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionsMenu.java [hunk 6]
```diff
@@ -103,13 +103,15 @@ public void fill(final Menu menu, int index) {
...
+			final @Nullable List<@Nullable Either<Command, CodeAction>> codeActions) {
 		final var job = new UIJob(menu.getDisplay(), Messages.updateCodeActions_menu) {
 			@Override
-			public IStatus runInUIThread(IProgressMonitor monitor) {
-				if (u != null) {
+			public IStatus runInUIThread(@Nullable IProgressMonitor monitor) {
+				if (ex != null) {
 					final var item = new MenuItem(menu, SWT.NONE, index);
-					item.setText(u.getMessage());
+					item.setText(String.valueOf(ex.getMessage()));
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionsMenu.java:114]: why do we add `String.valueOf`? Could we not leave it as it was?

**Comment by sebthom** [comment]
> @sebthom , you just asked for the review at the beginning of my summer holidays ;) I will check it today.

@rubenporras no worries, I hope you are well rested :)

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java [hunk 7]
```diff
@@ -289,17 +309,17 @@ public IContextInformation[] computeContextInformation(ITextViewer viewer, int o
...
+					.collectAll(ls -> ls.getTextDocumentService().signatureHelp(param).thenAccept(signatureHelp -> {
+						if (signatureHelp != null) {
+							signatureHelp.getSignatures().stream().map(LSContentAssistProcessor::toContextInformation)
+									.forEach(contextInformations::add);
+						}
+					}));
 			this.contextInformationLanguageServersFuture.get(CONTEXT_INFORMATION_TIMEOUT, TimeUnit.MILLISECONDS);
 		} catch (ResponseErrorException | ExecutionException e) {
-			if (!CancellationUtil.isRequestCancelledException(e)) { // do not report error if the server has cancelled the request
+			if (!CancellationUtil.isRequestCancelledException(e)) { // do not report error if the server has cancelled
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java:321]: the indentation of this comment is a bit weird now (at least on the GitHub review. Could we revert the change, or make it shorter so that the auto-format leaves it all in one line?

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java [hunk 8]
```diff
@@ -334,15 +355,18 @@ private void getFuture(CompletableFuture<@NonNull List<@NonNull Void>> future) {
 			LanguageServerPlugin.logError(e);
 			Thread.currentThread().interrupt();
 		} catch (TimeoutException e) {
-			LanguageServerPlugin.logWarning("Could not get trigger characters due to timeout after " + TRIGGERS_TIMEOUT + " milliseconds", e); //$NON-NLS-1$//$NON-NLS-2$
+			LanguageServerPlugin.logWarning(
+					"Could not get trigger characters due to timeout after " + TRIGGERS_TIMEOUT + " milliseconds", e); //$NON-NLS-1$//$NON-NLS-2$
 		} catch (OperationCanceledException | ResponseErrorException | ExecutionException | CancellationException e) {
-			if (!CancellationUtil.isRequestCancelledException(e)) { // do not report error if the server has cancelled the request
+			if (!CancellationUtil.isRequestCancelledException(e)) { // do not report error if the server has cancelled
+																	// the request
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java:362]: same comment about the format of the comment

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/linkedediting/LSPLinkedEditingReconcilingStrategy.java [hunk 9]
```diff
@@ -94,13 +96,13 @@ public void install(ITextViewer viewer) {
...
 			editorSelectionChangedListener = new EditorSelectionChangedListener();
-			editorSelectionChangedListener.install(sourceViewer.getSelectionProvider());
+			editorSelectionChangedListener.install(thisViewer.getSelectionProvider());
 		}
 	}
 
 	@Override
 	public void uninstall() {
-		if (sourceViewer != null) {
+		if (sourceViewer != null && editorSelectionChangedListener != null) {
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/linkedediting/LSPLinkedEditingReconcilingStrategy.java:105]: `EditorSelectionChangedListener` is still nullable, why do you remove the null check?

File: org.eclipse.lsp4e/.project [hunk 10]
```diff
@@ -21,12 +21,12 @@
 			</arguments>
 		</buildCommand>
 		<buildCommand>
-			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<name>org.eclipse.pde.ds.core.builder</name>
```
[the remark this answers is not in the document]
Comment (sebthom) [reply org.eclipse.lsp4e/.project:24]: We could but evertime I checkout or switch a branch with Eclipse, the entries get modified/reordered, so I guessed that is how new Eclipse or m2e versions like it.

File: org.eclipse.lsp4e/.settings/org.eclipse.jdt.core.prefs [hunk 11]
```diff
@@ -1,13 +1,17 @@
...
 org.eclipse.jdt.core.compiler.annotation.nonnull=org.eclipse.jdt.annotation.NonNull
 org.eclipse.jdt.core.compiler.annotation.nonnull.secondary=org.eclipse.lsp4j.jsonrpc.validation.NonNull
 org.eclipse.jdt.core.compiler.annotation.nonnullbydefault=org.eclipse.jdt.annotation.NonNullByDefault
 org.eclipse.jdt.core.compiler.annotation.nonnullbydefault.secondary=
+org.eclipse.jdt.core.compiler.annotation.notowning=org.eclipse.jdt.annotation.NotOwning
 org.eclipse.jdt.core.compiler.annotation.nullable=org.eclipse.jdt.annotation.Nullable
 org.eclipse.jdt.core.compiler.annotation.nullable.secondary=
 org.eclipse.jdt.core.compiler.annotation.nullanalysis=enabled
+org.eclipse.jdt.core.compiler.annotation.owning=org.eclipse.jdt.annotation.Owning
+org.eclipse.jdt.core.compiler.annotation.resourceanalysis=disabled
```
[the remark this answers is not in the document]
Comment (sebthom) [reply org.eclipse.lsp4e/.settings/org.eclipse.jdt.core.prefs:14]: that is the default setting. it is set by eclipse 2024-06 automatically.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionsMenu.java (hunk 6 above)
Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionsMenu.java:114]: get Message is nullable but setText expects a non null. value

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java (hunk 7 above)
Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java:321]: I could, but this is done by the current Eclipse formatter settings and I prefer to use the formatter settings because otherwise we have to manually update readjust the code everytime when we do code changes and use the formatter.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/linkedediting/LSPLinkedEditingReconcilingStrategy.java (hunk 9 above)
Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/linkedediting/LSPLinkedEditingReconcilingStrategy.java:105]: I didnt remove the null check, I added it.

Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/linkedediting/LSPLinkedEditingReconcilingStrategy.java:105]: I didnt remove the null check, I added it.

**Review by rubenporras — APPROVED** [review]

**Review by rubenporras — APPROVED** [review]

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java (hunk 7 above)
Comment (sebthom) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java:321]: I would actually really appreciate if we could harmonize the formatter settings of all LSP4E modules (they currently differ) and in a single commit reformat the whole project with these settings.

**Comment by sebthom** [comment]
@rubenporras thanks for the review and your trust!

**Comment by rubenporras** [comment]
Let cross fingers 👍
